design/toolchain.md

```diff
@@ -0,0 +1,115 @@
+
+## Current State
+
+CDK apps can be deployed today using the CDK CLI (or CDK Toolkit). The toolkit is implemented as a monolithic program with no clear boundaries between the various stages. We would like to break up the monolithic process executed by the toolkit in order to synthesize, package and deploy a CDK app.
+
+There a few reasons why we want to do this:
+
+1. __Security__: isolate the "deploy" step such that no user code needs to run. This is very important from a security perspective because deployment commonly require administrator privileges on the AWS account, and we need to reduce the attack surface during that time (see #2073, which currently has to run both build and deploy together in the same CodeBuild task).
+2. __Modularity__: Allow tools to utilize the various steps used to deploy a CDK app inside other tools such as IDEs, deployment tools, etc.
+3. __Code Quality__: the CLI's code base needs a fresh rewrite, along with complete unit test coverage and this is an opportunity to do that.
+
+## Requirements
+
+* It should be possible to execute each component in the toolchain individually by feeding it the output from the previous step.
+* Given a specific input, the output from each step must be completely reproducible (no side effects).
+* Different components may require different execution environments and/or permissions to run. For example `cdk-synth` may need to be able to query the target AWS account in order to resolve environmental context, `cdk-bundle` may need to build docker images, `cdk-deploy` will need admin permissions in order to deploy the app.
+* It should be possible to invoke each component as a jsii library from all supported languages.
+
+## Design
+
+Let's start by formally defining the states and steps in a CDK application's lifecycle:
+
+```
++-----+                +---------+             +----------+             +------------+
+| App |-- synthesis -->| Staging |-- bundle -->| Assembly |-- deploy -->| Deployment |
++-----+                +---------+             +----------+             +------------+
+```
+
+### cdk-synth
+
+We start from a ready-to-run (__compiled__) CDK app.
+
+Some languages require compilation (such as Java, Go, TypeScript) and in others (JavaScript, Ruby, Python), there is no need to compile. Compiling code is __not in scope__ for the CDK toolchain in order to allow developers to use their favorite IDEs and build toolchain idiomatically.
+
+The first component in the toolchain is called `cdk-synth`. It's purpose is to execute the CDK app so it can synthesize the CDK tree (through the internal app phases which are "init", "prepare", "validate", "synthesize"). Eventually the CDK tree can synthesize both final artifacts such as CloudFormation templates and intermediate artifacts such as unpackaged assets.
+
+The protocol between the CDK app `cdk-synth`:
+
+- Context is passed into the app via a JSON map serialized into an environment variable called `CONTEXT_ENV`.
```

Comment:
Can't help feel those are clumsy. Also, all environment variables used by the CDK should have names starting with `CDK_` in my opinion.